pysyncobj/journal.py

```diff
@@ -138,32 +258,78 @@ def __init__(self, journalFile):
+        self.__flushJournal = flushJournal
 
     def __getDefaultHeader(self):
         appName = APP_NAME + b'\0' * (NAME_SIZE - len(APP_NAME))
         appVersion = APP_VERSION + b'\0' * (VERSION_SIZE - len(APP_VERSION))
-        header = appName + appVersion + struct.pack('<II', JOURNAL_FORMAT_VERSION, FIRST_RECORD_OFFSET)
+        header = (appName + appVersion + struct.pack('<I', JOURNAL_FORMAT_VERSION) +
+          struct.pack('<Q', 0) + # default term = 0
+          VOTED_FOR_NONE_HASH + # default voted for = empty
+          struct.pack('<I', FIRST_RECORD_OFFSET))
         return header
 
     def __getLastRecordOffset(self):
         return struct.unpack('<I', self.__journalFile.read(LAST_RECORD_OFFSET_OFFSET, 4))[0]
 
     def __setLastRecordOffset(self, offset):
         self.__journalFile.write(LAST_RECORD_OFFSET_OFFSET, struct.pack('<I', offset))
-
-    def add(self, command, idx, term):
+        # No auto-flushing needed here because it's called in the methods below.
+
+    @property
+    def currentTerm(self):
+        return self.__currentTerm
+
+    @currentTerm.setter
+    def currentTerm(self, term):
+        self.__set_currentTerm(term)
+
+    @property
+    def votedForNodeId(self):
+        return self.__votedForNodeIdProxy
+
+    @votedForNodeId.setter
+    def votedForNodeId(self, nodeId):
+        self.__set_votedForNodeId(nodeId)
+
+    def __set_currentTerm(self, term, flush = True):
+        self.__journalFile.write(NAME_SIZE + VERSION_SIZE + 4, struct.pack('<Q', term))
+        if flush and self.__flushJournal:
+            self.flush()
+        self.__currentTerm = term
+
+    def __set_votedForNodeId(self, nodeId, flush = True):
+        self.__votedForNodeIdHash = hashlib.md5(pysyncobj.pickle.dumps(nodeId)).digest()
+        self.__journalFile.write(NAME_SIZE + VERSION_SIZE + 4 + CURRENT_TERM_SIZE, self.__votedForNodeIdHash)
+        if flush and self.__flushJournal:
+            self.flush()
+        if self.__votedForNodeIdHash != VOTED_FOR_NONE_HASH:
+            self.__votedForNodeIdProxy = VotedForNodeIdHashProxy(_hash = self.__votedForNodeIdHash)
+        else:
+            self.__votedForNodeIdProxy = None
+
+    def set_currentTerm_and_votedForNodeId(self, term, nodeId):
+        # Only flush once
+        self.__set_currentTerm(term, flush = False)
+        self.__set_votedForNodeId(nodeId, flush = True)
+
+    def add(self, command, idx, term, _doFlush = True):
         self.__journal.append((command, idx, term))
-        cmdData = struct.pack('<QQ', idx, term) + to_bytes(command)
+        cmdData = struct.pack('<QQ', idx, term) + pysyncobj.pickle.to_bytes(command)
         cmdLenData = struct.pack('<I', len(cmdData))
         cmdData = cmdLenData + cmdData + cmdLenData
         self.__journalFile.write(self.__currentOffset, cmdData)
         self.__currentOffset += len(cmdData)
         self.__setLastRecordOffset(self.__currentOffset)
+        if _doFlush and self.__flushJournal:
```

Comment:
I suppose, flushing every transaction can heart performance. I suggest to do following:
1) write some benchmark (for journal standalone) to check how much performance degradates
2) may be somehow add __getLastFlushedLogTerm in addition to __getCurrentLogTerm and call flush in async mode (or at least once per second?)

Reply:
ok, just please check performance to ensure that will not drops dramatically

Reply:
(Moving the performance discussion to the PR level since it's not really specific to this line in journal.py.)